packages/fluentui/react-builder/src/narration/DescendantsNarrationsComputer.tsx

```diff
@@ -0,0 +1,80 @@
+import { NarrationComputer, IAriaElement } from './NarrationComputer';
+
+interface INarration {
+  path: string[];
+  text: string;
+}
+
+export class DescendantsNarrationsComputer {
+  computer: NarrationComputer = new NarrationComputer();
+
+  // Computes the screen reader narrations for the given element and all its focusable descendants, for the given platform.
+  async compute(element: IAriaElement, platform: string): Promise<INarration[]> {
+    // Prepare all the arrays
+    const activeDescendantsParents: IAriaElement[] = [];
+    const path: string[] = [];
+    const narrations: INarration[] = [];
+
+    // Traverse down the DOM tree rooted at the element and compute the activeDescendantsParents, path and narrations along the way
+    this.findActiveDescendantsParents(element, activeDescendantsParents);
+    await this.traverse(element, platform, path, narrations, activeDescendantsParents);
+
+    return narrations;
+  } // End compute
+
+  // Traverses the DOM tree rooted at the given element to find all the parents which have the ariaActiveDescendantElement property set, and saves the found elements to the given parents array.
+  findActiveDescendantsParents(element: IAriaElement, parents: IAriaElement[]) {
+    if (element.ariaActiveDescendantElement != null) {
+      // Begin if 1
+      parents.push(element);
+    } // End if 1
+    for (let i = 0; i < element.children.length; i++) {
+      // Begin for 1
+      const child = element.children[i] as IAriaElement;
+      this.findActiveDescendantsParents(child, parents);
+    } // End for 1
+  } // End findActiveDescendantsParents
+
+  // Traverses the given element and computes the narrations for it and its descendants which are focusable..
+  async traverse(
+    element: IAriaElement,
+    platform: string,
+    path: string[],
+    narrations: INarration[],
+    activeDescendantsParents: IAriaElement[],
+  ) {
+    // Determine the path
+    const newPath = path.slice();
+    const node = await (window as any).getComputedAccessibleNode(element);
+
+    // Only add the item  to the path if the role is not a generic container
+    if (node.role !== 'genericContainer') {
+      const item = node.name ? `${node.role} (${node.name})` : node.role;
+      newPath.push(item);
+    }
+
+    // If the element is focusable, stop the traversal and compute and save the narration
+    const isDirectlyFocusable =
+      (element.getAttribute('tabindex') || element.tabIndex >= 0) && element.ariaActiveDescendantElement == null;
+    const isActiveDescendant = activeDescendantsParents.some(parent => {
+      // Begin some 1
+      return parent.tabIndex >= 0 && parent.ariaActiveDescendantElement === element;
+    }); // End some 1
```

Comment:
Why is this check necessary ? if you are traversing the tree wouldn't you find the directly focusable parent first ?

Reply:
Sorry, I didn't get it. Can you explain on it more?

Reply:
Ah sorry, I might not have understood the logic completely so this comment might be unnecessary.

If you were traversing the tree and found an element that was focusable you would stop traversing. Therefore do you need to check if an element has focusable parents ?

Reply:
As discussed offline, I didn't quite understand that you have to keep traversing the if there is an active descendant